File: 0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        if (left == right) return head;

        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        head = dummy;

        ListNode* preLeft = head;

        for (int i = 1; i < left; i++) {
            preLeft = preLeft->next;
        }

        ListNode* cur = preLeft->next;
        ListNode* prev = nullptr;
        ListNode* nextNode = nullptr;

        for (int i = left; i <= right; i++) {
            nextNode = cur->next;
            cur->next = prev;
            prev = cur;
            cur = nextNode;
        }

        preLeft->next->next = cur;
        preLeft->next = prev;

        return dummy->next;
    }
};

File: 0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        if left == right: return head
        dummy = ListNode(0)
        dummy.next = head
        head = dummy
        
        preLeft = head
        i = 1
        while i < left:
            preLeft = preLeft.next
            i += 1
        cur = preLeft.next
        prv = None
        nextNode = None
        
        while i <= right:
            nextNode = cur.next
            cur.next = prv
            prv = cur
            cur = nextNode
            i+=1
        
        preLeft.next.next = cur
        preLeft.next = prv
        return head.next
        